Implement binding and lowering for the with expression

agocke opened the pull request:
A with expression binds to a parameterless, accessible, instance method
called Clone on the receiver type.

Then, all arguments are bound against record properties with the same
name, and the expressions are assigned to the properties' backing
fields.

Relates to https://github.com/dotnet/roslyn/issues/40726 (test plan for records)

gafter commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -6046,6 +6046,18 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_ExpressionTreeContainsRangeExpression" xml:space="preserve">
     <value>An expression tree may not contain a range ('..') expression.</value>
   </data>
+  <data name="WRN_GeneratorFailedDuringGeneration" xml:space="preserve">
```
>WRN_GeneratorFailedDuringGeneration [](start = 14, length = 35)

This doesn't look right.

gafter commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -6055,16 +6067,16 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
-  <data name="WRN_GeneratorFailedDuringGeneration" xml:space="preserve">
-    <value>Generator '{0}' failed to generate source. It will not contribute to the output and compilation errors may occur as a result.</value>
+  <data name="ERR_InvalidWithReceiverType" xml:space="preserve">
+    <value>The receiver of a `with` expression must have a valid non-void type.</value>
```
>valid [](start = 59, length = 5)

What does "valid" mean here?  I suggest dropping the word.  If it were an error type we presumably would not report this.

gafter commented:
There are a handful of parts of the specification that don't make sense to me, but I don't know where to put comments on it. I can't really review the implementation against a spec that I don't understand.

agocke commented:
I created a spec PR that we can use for discussion here: https://github.com/dotnet/csharplang/pull/3353

333fred commented on src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs:
```diff
@@ -0,0 +1,183 @@
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// This portion of the binder converts an <see cref="ExpressionSyntax"/> into a <see cref="BoundExpression"/>.
```
Seems like it converts a `WithExpressionSyntax` specifically.

gafter replied:
I don't think doc comments on parts of a partial "work" properly.  This could be an ordinary non-doc comment.

---
In reply to: [413343653](https://github.com/dotnet/roslyn/pull/43249#discussion_r413343653) [](ancestors = 413343653)

333fred commented on src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs:
```diff
@@ -0,0 +1,183 @@
+                        {
+                            case SymbolKind.Property:
+                                // PROTOTYPE: this should check for init-only, but that isn't a separate feature yet
+                                if (!(sym is SynthesizedRecordPropertySymbol))
```
This currently means we only support records in the same compilation, correct? Just making sure I understand the current limitations

agocke replied:
That's correct.

333fred commented on src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs:
```diff
@@ -0,0 +1,183 @@
+                                // PROTOTYPE: this should check for init-only, but that isn't a separate feature yet
+                                if (!(sym is SynthesizedRecordPropertySymbol))
+                                {
+                                    hasErrors = true;
```
Consider just `goto default`

333fred commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -426,5 +432,929 @@ public void EmptyRecord()
+}";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (7,13): error CS8803: The 'with' expression requires the receiver type 'C' to have a single accessible non-inherited instance method named "Clone".
```
Not generating clone by default currently? If that's the point of this test, consider using a non-record type, as we'll just have to adjust it when we start generating them. And if the point is to have a place that will change when we do start generating, consider leaving a prototype comment to record that intention.

agocke replied:
Yup, just didn't want to deal with both parts at the same time since they should be somewhat independent.

333fred commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -426,5 +432,929 @@ public void EmptyRecord()
+}";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (12,13): error CS8803: The 'with' expression requires the receiver type 'B' to have a single accessible non-inherited instance method named "With".
```
>With [](start = 160, length = 4)

Clone?

333fred replied:
And a few others

---
In reply to: [413365604](https://github.com/dotnet/roslyn/pull/43249#discussion_r413365604) [](ancestors = 413365604)

333fred commented:
Done review pass (commit 2)

333fred commented on src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs:
```diff
@@ -57,10 +57,7 @@ private BoundExpression BindWithExpression(WithExpressionSyntax syntax, Diagnost
-                                    diagnostics.Add(
-                                        ErrorCode.ERR_WithMemberIsNotRecordProperty,
-                                        location);
+                                    goto default;
```
This is actually a behavior change: before we'd error and set `member = sym`, now we don't set `member = sym`. That's why I was suggesting doing it after that part. If there's no test that demonstrates this, consider adding one.

agocke replied:
Oops, good catch. I'll verify

gafter commented on src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs:
```diff
@@ -0,0 +1,180 @@
+                            case SymbolKind.Property:
+                                member = sym;
+                                // PROTOTYPE: this should check for init-only, but that isn't a separate feature yet
+                                if (!(sym is SynthesizedRecordPropertySymbol))
```
>SynthesizedRecordPropertySymbol [](start = 45, length = 31)

Please add a PROTOTYPE comment saying this does not work from metadata.

gafter commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_WithExpression.cs:
```diff
@@ -0,0 +1,61 @@
+                var prop = (SynthesizedRecordPropertySymbol)arg.Member;
+                stores.Add(F.AssignmentExpression(
+                    (BoundExpression)F.Field((BoundExpression)receiverLocal, (FieldSymbol)prop.BackingField),
+                    (BoundExpression)VisitExpression((BoundExpression)arg.Expression)
```
This is not correct; the type of the receiver (base record) might not contain that property (a property of the derived receiver type).

gafter commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_WithExpression.cs:
```diff
@@ -0,0 +1,61 @@
+            foreach (var arg in withExpr.Arguments)
+            {
+                RoslynDebug.AssertNotNull(arg.Member);
+                var prop = (SynthesizedRecordPropertySymbol)arg.Member;
```
>SynthesizedRecordPropertySymbol [](start = 28, length = 31)

Please add a PROTOTYPE comment that this will only work for source symbols.

gafter commented on src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs:
```diff
@@ -0,0 +1,180 @@
+                cloneMethod,
+                args.ToImmutableAndFree(),
+                cloneMethod?.ReturnType ?? receiverType,
+                hasErrors: hasErrors);
```
I cannot tell if this is correct as the specification does not tell me what the type of a with expression is.

gafter commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -426,5 +432,955 @@ public void EmptyRecord()
+}";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (12,13): error CS8803: The 'with' expression requires the receiver type 'C' to have a single accessible non-inherited instance method named "Clone".
```
>non-inherited instance [](start = 123, length = 22)

This is not the diagnostic in the resources.  Please update.

gafter commented on src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs:
```diff
@@ -0,0 +1,180 @@
+            }
+
+            // PROTOTYPE: The receiver type must have a instance method called 'Clone' with no parameters
+            LookupMembersWithoutInheritance(
```
>WithoutInheritance [](start = 25, length = 18)

"Without Inheritance" is not justified by the specification.  Should the spec say "The receiver type must **declare** ..." instead of "The receiver type must **contain** ..."?

gafter commented on src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs:
```diff
@@ -0,0 +1,180 @@
+                    var location = initializer.NameEquals!.Name.Location;
+                    this.LookupMembersInType(
+                        lookupResult,
+                        receiverType,
```
>receiverType [](start = 24, length = 12)

You are looking in the wrong type.  According to the spec we are to look in the type of the return of the `Clone` method.

gafter commented on src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs:
```diff
@@ -0,0 +1,180 @@
+            }
+
+            lookupResult.Clear();
+            useSiteDiagnostics = null;
```
>useSiteDiagnostics = null [](start = 12, length = 25)

Please add a PROTOTYPE comment that use site diagnostics are lost.

gafter commented on src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs:
```diff
@@ -0,0 +1,180 @@
+                        cloneMethod.ReturnType);
+                }
+
+                useSiteDiagnostics = null;
```
>useSiteDiagnostics = null [](start = 16, length = 25)

Please add a PROTOTYPE comment that use site diagnostics are lost.

gafter commented:
Finished reviewing Iteration 4.

agocke commented:
@gafter I think I've addressed your comments

gafter commented on src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs:
```diff
@@ -0,0 +1,175 @@
+                LookupMembersInType(
+                    lookupResult,
+                    receiverType,
+                    "Clone",
```
>"Clone" [](start = 20, length = 7)

We tend to use constants for well known names in "WellKnownMemberNames".

gafter commented on src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs:
```diff
@@ -0,0 +1,175 @@
+                        if (symbol is MethodSymbol { ParameterCount: 0 } m)
+                        {
+                            cloneMethod = m;
+                            break;
```
How does this handle multiple results? That can happen for example when `Clone` is overridden. This technique does not correspond to how invocations are handled, as there is no contract for the order of symbols in a lookup result.

gafter commented on src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs:
```diff
@@ -0,0 +1,175 @@
+                        originalBinder: this,
+                        diagnose: false,
+                        useSiteDiagnostics: ref useSiteDiagnostics);
+                    if (lookupResult.IsSingleViable &&
```
>IsSingleViable [](start = 37, length = 14)

How does this handle properties that are overridden or hidden (more than one symbol in the lookup result)?  This does not correspond to how expressions are bound.

gafter commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -604,5 +609,1000 @@ class C
+            var comp = CreateCompilation(src);
+            // PROTOTYPE: records don't auto-generate Clone at the moment
+            comp.VerifyDiagnostics(
+                // (7,13): error CS8803: The 'with' expression requires the receiver type 'C' to have a single accessible non-inherited instance method named "Clone".
```
>non-inherited [](start = 122, length = 13)

This is not the diagnostic in the resources.  Please update.

gafter commented:
Finished reviewing (Iteration 6). I think there are four active comments from me.

agocke commented:
@gafter Thanks, updated. Regarding the lookup that will be adjusted soon in PRs unifying init-only and generating record Clone, so I've added prototype comments so I can include that in later PRs.
